common/include/RevCommon.h

```diff
@@ -45,6 +45,21 @@ constexpr auto SignExt(T val, size_t bits){
   return static_cast<std::make_signed_t<T>>((ZeroExt(val, bits) ^ signbit) - signbit);
 }
 
+/// Base-2 logarithm of integers
+template<typename T>
+constexpr int lg(T x){
+  static_assert(std::is_integral_v<T>);
+
+  // We select the __builtin_clz which takes integers no smaller than x
+  if constexpr(sizeof(x) <= sizeof(int)){
+    return x ? 8*sizeof(int)-1 - __builtin_clz(x) : -1;
+  }else if constexpr(sizeof(x) <= sizeof(long)){
+    return x ? 8*sizeof(long)-1 - __builtin_clzl(x) : -1;
+  }else{
+    return x ? 8*sizeof(long long)-1 - __builtin_clzll(x) : -1;
```

Comment:
should the case of x == 0 throw rather than return -1? it's a sort-of-NaN.

Reply:
We don't want to return `0` because that coincides with the valid case of `lg(1) = 0`. Returning `-1` seems to be the best way of indicating error. We don't need to throw exceptions or call error routines -- this is supposed to be a simple helper function. Also it's `constexpr`, and by implication, `inline` (when I tried a simple test case, the constant was propagated and the function was fully evaluated at compile-time). The caller can check the argument and raise errors if the argument is invalid.

The x86 `bsf` and `bsr` instructions (as well as `__builtin_ctz()`) produce undefined behavior if the input is `0`, which is why we test for it.

Also see [Find first set](https://en.wikipedia.org/wiki/Find_first_set).